src/Bicep.Core/Emit/InlineDependencyVisitor.cs

```diff
@@ -198,29 +198,44 @@ private void VisitPropertyAccessSyntaxInternal(PropertyAccessSyntax syntax)
                 return;
             }
 
-            static bool ShouldSkipInlining(ObjectType objectType, string propertyName)
+            static bool ShouldSkipInlining(ObjectType objectType, string propertyName, ResourceDeclarationSyntax? resourceSyntax = null)
             {
                 if (!objectType.Properties.TryGetValue(propertyName, out var propertyType))
                 {
                     // unknown property - we should assume it's not inlinable
                     return true;
                 }
 
-                // update the cache if property can't be skipped for inlining
-                return propertyType.Flags.HasFlag(TypePropertyFlags.DeployTimeConstant);
+                if (propertyType.Flags.HasFlag(TypePropertyFlags.DeployTimeConstant))
+                {
+                    if (resourceSyntax is not null &&
+                        propertyName != LanguageConstants.ResourceIdPropertyName &&
+                        propertyName != LanguageConstants.ResourceNamePropertyName &&
+                        propertyName != LanguageConstants.ResourceTypePropertyName &&
+                        propertyName != LanguageConstants.ResourceApiVersionPropertyName &&
+                        resourceSyntax.TryGetBody()?.SafeGetPropertyByName(propertyName) is null)
+                    {
+                        // The property is not declared in the resource - we should inline it even it is a deploy-time constant.
+                        return false;
+                    }
+
+                    return true;
+                }
+
+                return false;
             }
 
             switch (model.GetSymbolInfo(variableAccessSyntax))
             {
                 // Note - there's a limitation here that we're using the 'declared' type and not the 'assigned' type.
                 // This means that we may encounter a DiscriminatedObjectType. For now we should accept this limitation,
                 // and move to using the assigned type once https://github.com/Azure/bicep/issues/1177 is fixed.
-                case ResourceSymbol { Type: ResourceType { Body: { Type: ObjectType bodyObjectType } } }:
-                    SetCache(!ShouldSkipInlining(bodyObjectType, syntax.PropertyName.IdentifierName));
+                case ResourceSymbol { Type: ResourceType { Body: { Type: ObjectType bodyObjectType } }, DeclaringResource: { } resourceSyntax }:
```

Comment:
There are quite a few places now where we have this logic to get the body type regardless of whether we have a single resource or a collection. Worth commonizing that logic?

E.g.
https://github.com/Azure/bicep/blob/85842563b94d63035c6086234b39b25aa245677f/src/Bicep.Core/Emit/ScopeHelper.cs#L388-L393

https://github.com/Azure/bicep/blob/85842563b94d63035c6086234b39b25aa245677f/src/Bicep.Core/Emit/ScopeHelper.cs#L525-L530

Reply:
Turns out the logic is not just used in `Emit`, so I added the helper methods to `ResourceSymbol` and `ModuleSymbol`.